frame/vault/src/tests.rs

```diff
@@ -522,96 +532,100 @@ proptest! {
+			// BEFORE losses and profits
+			for (account, initial_native_tokens, _) in before_moment_lps() {
+				let origin = Origin::signed(account);
+				prop_assert_ok!(Vaults::deposit(origin, vault_id, initial_native_tokens));
 			}
 
-			// Profits comming
-			prop_assert_ok!(
-				<Vaults as StrategicVault>::deposit(
-					&vault_id,
-					&strategy_account_id,
-					strategy_profits
-				)
-			);
-
-			// Shareholders total LP
-			let lp_total = Tokens::total_issuance(vault.lp_token_id);
+			prop_assert_ok!(<Vaults as StrategicVault>::deposit(
+				&vault_id,
+				&strategy_account_id,
+				strategy_profits
+			));
+			prop_assert_ok!(<Vaults as StrategicVault>::withdraw(
+				&vault_id,
+				&strategy_account_id,
+				strategy_losses
+			));
+			let lp_tokens_total = Tokens::total_issuance(vault.lp_token_id);
 
-			// Users depositing later
-			for (account, balance) in accounts[strategy_deposit_moment..accounts.len()].iter().copied() {
-				prop_assert_ok!(
-					Vaults::deposit(Origin::signed(account), vault_id, balance)
-				);
+			// Liquidity providers deposit all their native tokens to receive LP tokens
+			// AFTER losses and profits
+			for (account, initial_native_tokens, _) in after_moment_lps() {
+				let origin = Origin::signed(account);
+				prop_assert_ok!(Vaults::deposit(origin, vault_id, initial_native_tokens));
 			}
 
-			// Withdraw & local check
-			for ((account, balance), index) in accounts.iter().copied().zip(account_start..) {
-				// Current lp
-				let lp = Tokens::balance(vault.lp_token_id, &account);
+			// Withdraw & native tokens checks
+			for (idx, (account, initial_native_tokens, _)) in lps.iter().copied().enumerate() {
+				let lp_tokens = Tokens::balance(vault.lp_token_id, &account);
 
-				// Withdraw all my shares, including profits
-				prop_assert_ok!(Vaults::withdraw(Origin::signed(account), vault_id, lp));
+				// Withdraws all LP tokens
+				prop_assert_ok!(Vaults::withdraw(Origin::signed(account), vault_id, lp_tokens));
+				// New balance that includes losses and profits
+				let new_native_tokens = Tokens::balance(asset_id, &account);
 
-				// Balance after having deposited + withdrawn my funds
-				let new_balance = Tokens::balance(asset_id, &account);
+				let curr_lp_deposited_before_moment = lps_start + idx <= strategy_moment;
 
-				// We had shares before the profit, we get a cut of the profit
-				if index <= strategy_deposit_moment {
-					// Compute my share
-					let strategy_profit_share =
-						multiply_by_rational(strategy_profits, lp, lp_total).expect("qed;");
+				if curr_lp_deposited_before_moment {
+					let strategy = multiply_by_rational(
+						strategy_profits,
+						lp_tokens,
+						lp_tokens_total,
+					)
+					.expect("qed;");
 
-					prop_assert_epsilon!(new_balance, balance + strategy_profit_share);
-				}
-				else {
+					prop_assert_epsilon!(new_native_tokens, initial_native_tokens + strategy);
```

Comment:
Quite didn't understand when withdraw should be called. Within strategy moment? After profits checks? Create another strategy moment for withdraws?

Since the unmodified `vault_stock_dilution_k` test checks profits of a full LP tokens withdraw, I tried to make the strategy deposit and then withdraw with the amount of deposit greater than the amount of withdraw. It would then be possible to assert a small amount of profits that are influenced by the withdrawn amount.

Reply:
There is many scenarios which are untested... Basically the one you describe but many others. For instance, since the vault has a reserve, this `stock_dilution_k` would not work if the strategy was withdrawing, because the vault would not own 100% of the funds. So it would also be nice to check for "incremental" withdrawing from users. Like the strategy would deposit each X blocks (increasing vault funds), allowing users to withdraw.